### PR TITLE
Analyze large datums

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
Mask columns exceeding threshold in analyze sample

    Analyze collects a sample from the table, in case the
    sample contains columns with huge length, the columns may
    result memory usage to go high cancelling the query.

    This commit masks the lengthy columns to avoid high
    memory usage while collecting sample. Column values exceeding
    WIDTH_THRESHOLD will be ignored from the samples.

    Approach to address this issue :

     Numeric types or Fixed length user defined types:
      Always include the column in the sample.

    Variable length types:
      Check the toasted/compressed size and mask the column value if size
      exceeds column_threshhold (compressed)

### CODE REVIEW

**Comment by bhuvnesh2703** [comment]
@hlinnaka Continued further on it following the discussion on https://github.com/greenplum-db/gpdb/pull/2073/. 
Started with the approach to encapsulate the changes into std_fetch_func, however realized that std_fetch_func is called from 
compute_scalar_stats, and if we have to pass the array of offsets into it, it appears to be by changing the call for compute_scalar_stats 
and pass the array to it so that std_fetch_func can call it `value = fetchfunc(stats, i, &isnull); in compute_scalar_stats` as for every column the array offset can be different. 
Other option could have been to add one more attribute to VacAttrStats which will hold the offset or use globals but avoided them.
So, opted to populate stats->rows to contain only valid rows for that column and passed the number of valid rows to compute scalar stats. 
This filters the bad rows before calling compute_scalar_stats, however iterates through the rows array to construct a new array if there is any large datum. 

In case of indexes, the tuples are converted to a array [Row1Col1Val, Row1Col2Val, Row2Col1Val, Row2Col2Val] and in ind_fetch_func
based on the attribute count and rowno we get the expr values and again as it is called from compute_scalar_stats, found limited options to change, so while evaluating the exprvals, marked the bad rows with Datum (0) and isNull false, and filtered them further in compute_scalar_stats. I hope that change in compute_stats is trivial.
Please let me know for any concerns with this approach. Incorporated the other comments.

**Comment by hlinnaka** [comment]
On 04/04/2017 10:07 AM, Bhuvnesh wrote:
> @hlinnaka Continued further on it following the discussion on https://github.com/greenplum-db/gpdb/pull/2073/.
> Started with the approach to encapsulate the changes into std_fetch_func, however realized that std_fetch_func is called from
> compute_scalar_stats, and if we have to pass the array of offsets into it, it appears to be by changing the call for compute_scalar_stats
> and pass the array to it so that std_fetch_func can call it `value = fetchfunc(stats, i, &isnull); in compute_scalar_stats` as for every column the array offset can be different.
> Other option could have been to add one more attribute to VacAttrStats which will hold the offset or use globals but avoided them.
> So, opted to populate stats->rows to contain only valid rows for that column and passed the number of valid rows to compute scalar stats.
> This filters the bad rows before calling compute_scalar_stats, however iterates through the rows array to construct a new array if there is any large datum.

Ok, makes sense.

- Heikki

**Comment by hlinnaka** [comment]
Oh, I just left review comments on the old PR, I didn't notice you opened a new one..

**Comment by bhuvnesh2703** [comment]
@hlinnaka Thanks no worries, will check those..

File: src/backend/commands/analyze.c [hunk 1]
```diff
@@ -392,7 +402,7 @@ analyze_rel_internal(Oid relid, VacuumStmt *vacstmt,
 			}
 		}
 	}
-
+	RowIndexes colLargeRowIndexes[attr_cnt];
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: Allocating a variable-size array in stack isn't legal in standard C. At least not in all standards and versions :-). Use palloc().

File: src/backend/commands/analyze.c [hunk 2]
```diff
@@ -448,12 +459,33 @@ analyze_rel_internal(Oid relid, VacuumStmt *vacstmt,
 		for (i = 0; i < attr_cnt; i++)
 		{
 			VacAttrStats *stats = vacattrstats[i];
+			RowIndexes rowIndexes = colLargeRowIndexes[i];
+			int validRowsLength = numrows - rowIndexes.toowide_cnt;
+			int validRowsIdx = 0;
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: You could move validRowsIdex to inside the if-block, it's only used in there.

File: src/backend/commands/analyze.c [hunk 3]
```diff
@@ -448,12 +459,33 @@ analyze_rel_internal(Oid relid, VacuumStmt *vacstmt,
...
+					if (rowIndexes.rows[rownum]) // If row is too wide, ignore it from sample
+						continue;
+					validRows[validRowsIdx] = rows[rownum];
+					validRowsIdx++;
+				}
+				stats->rows = validRows;
+			}
+			else
+			{
+				stats->rows = rows;
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: for clarity, I think it'd be good to also do `validRowsLength = numrows` here. And `validRowsLength = validRowsIdx` in the then-branch

File: src/backend/commands/analyze.c [hunk 4]
```diff
@@ -704,7 +742,16 @@ compute_index_stats(Relation onerel, double totalrows,
 					VacAttrStats *stats = thisdata->vacattrstats[i];
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: Hmm, this needs some re-thinking (and much better comments!).

At first, I was going to say that we should use the same mechanism here, to leave out those rows from the sample that had wide values, before calling compute_stats(). But that makes no sense. We're not computing stats here for a single collumn, but for an expression/predicate that can refer to any of the columns. For example, you might have an index on the expression "(colA + colB)". So, we cannot use the same mechanism with expressions or predicate indexes, 

Issues with the current code:

* The too-wide values are passed to the ExecQual to evaluate the predicate on partial indexes. Did we set them to NULL? I guess that means that any too-wide values will be treated as NULLs for the purposes of checking whether they match partial indexes. That's probably all right, I don't think we can do much better than that, and it's exceedingly rare to have an index on a column that's very wide, let alone a partial index. But it really needs a comment to explain that!

* I don't think the indexLargeRows array will work here. If I understand this correctly, the idea is that indexLargeRows[indexcol].rows[rowno] is set, if any of the table columns referenced by the index expression were too wide on this row. For example, if you have an index on `colA + colB`, then indexLargeRows[0].rows[0] would be set, if the first row in the sample had a too-wide column in either colA or colB. But I don't see any processing here to check all the relevant columns, so I don't think this is right.

Comment (hlinnaka) [inline src/backend/commands/analyze.c]: All in all, I'm starting to feel that this is too complicated for expression/partial indexes to bother. How about we leave compute_index_stats() unchanged, and just add a comment in the caller that we treat any too-wide datums as NULLs, for the purposes of expression/predicate index statistics? As I said above, expression/partial indexes are quite rare compared to normal ones, and it's even more rare to have one one very wide column.

File: src/backend/commands/analyze.c [hunk 5]
```diff
@@ -1715,7 +1816,6 @@ update_attstats(Oid relid, int natts, VacAttrStats **vacattrstats)
 		/* Ignore attr if we weren't able to collect stats */
 		if (!stats->stats_valid)
 			continue;
-
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: spurious whitespace change

File: src/include/commands/vacuum.h [hunk 6]
```diff
@@ -55,6 +55,17 @@
  */
 typedef struct VacAttrStats *VacAttrStatsP;
 
+/*
+ * Maintain the row index for large datums which must not be considered for
+ * samples while calculating statistcs
+ */
+typedef struct RowIndexes
```
Comment (hlinnaka) [inline src/include/commands/vacuum.h]: This is only used within analyze.c, would be better to move it there to not expose it to everyone.

File: src/test/regress/expected/analyze.out [hunk 7]
```diff
@@ -791,6 +791,26 @@ select * from pg_stats where tablename like 'p3_sales%' order by tablename, attn
...
...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa', 'bbbbb2', 'cccc2', 3);
```
Comment (hlinnaka) [inline src/test/regress/expected/analyze.out]: I'd suggest using repeat('a', 3640) here to make this less verbose.

**Comment by hlinnaka** [comment]
Ok, I copied my review comments from the old PR to here. I see that you had already addressed some of them.

Overall, I think we're making progress here, but still needs some work.

**Comment by bhuvnesh2703** [comment]
Thanks @hlinnaka for the feedback, incorporated the changes.. Again thank you for guiding up on the PR :). Ok makes sense to remove the changes for index statistics.. 
Made the changes tested using the regress analyze test, it passes. Let me know if there is any comment

File: src/backend/commands/analyze.c [hunk 8]
```diff
@@ -1389,13 +1441,44 @@ acquire_sample_rows_by_query(Relation onerel, int nattrs, VacAttrStats **attrsta
 	{
 		for (i = 0; i < nattrs; i++)
 		{
-			if (i != 0)
+			isVarlenaCol[i] = false;
+			NameData attname = attrstats[i]->attr->attname;
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: Might want to do `char       *attname = NameStr(attrstats[i]->attr->attname)` here, as all the uses of the variable pass it to NameStr anyway. Or even `quote_identifier(NameStr(attrstats[i]->attr->attname)`

File: src/backend/commands/analyze.c [hunk 9]
```diff
@@ -1342,7 +1393,7 @@ compare_rows(const void *a, const void *b)
 static int
```
Comment (hlinnaka) [inline src/backend/commands/analyze.c]: Please explain the new colLargeRowIndexes argument in the function comment.

**Comment by bhuvnesh2703** [comment]
@hlinnaka Thank you.. Sure, will add the test case for index stats and fix the other comments.

File: src/backend/commands/analyze.c [hunk 10]
```diff
@@ -448,16 +475,45 @@ analyze_rel_internal(Oid relid, VacuumStmt *vacstmt,
 		for (i = 0; i < attr_cnt; i++)
 		{
 			VacAttrStats *stats = vacattrstats[i];
+			RowIndexes *rowIndexes = colLargeRowIndexes[i];
+			int validRowsLength = numrows - rowIndexes->toowide_cnt;
+
+			/* If there are toowide rows in the sample, remove them
```
Comment (vraghavan78) [inline src/backend/commands/analyze.c]: too wide

File: src/backend/commands/analyze.c [hunk 11]
```diff
@@ -448,16 +475,45 @@ analyze_rel_internal(Oid relid, VacuumStmt *vacstmt,
...
+			 * from the sample being sent for stats collection
+			 */
+			if (rowIndexes->toowide_cnt > 0)
+				{
+				int validRowsIdx = 0;
+				for (int rownum=0; rownum < numrows; rownum++)
+				{
+					if (rowIndexes->rows[rownum]) // If row is too wide, ignore it from sample
```
Comment (vraghavan78) [inline src/backend/commands/analyze.c]: the sample

File: src/backend/commands/analyze.c [hunk 12]
```diff
@@ -448,16 +475,45 @@ analyze_rel_internal(Oid relid, VacuumStmt *vacstmt,
...
+				stats->rows = rows;
+				validRowsLength = numrows;
+			}
 
-			stats->rows = rows;
 			stats->tupDesc = onerel->rd_att;
 			(*stats->compute_stats) (stats,
 									 std_fetch_func,
-									 numrows,
+									 validRowsLength, // Numbers of rows in sample excluding toowide if any.
```
Comment (vraghavan78) [inline src/backend/commands/analyze.c]: Numbers -> numbers

File: src/backend/commands/analyze.c [hunk 13]
```diff
@@ -1389,13 +1446,44 @@ acquire_sample_rows_by_query(Relation onerel, int nattrs, VacAttrStats **attrsta
...
+			if (is_varlena || is_varwidth)
+			{
+				appendStringInfo(&columnStr,
+								 "(case when pg_column_size(Ta.%s) > %d then NULL else Ta.%s  end) as %s, ",
+								 attname,
+								 WIDTH_THRESHOLD,
+								 attname,
+								 attname);
+				appendStringInfo(&columnStr,
+								 "(case when Ta.%s is NULL then %s else %s end)",
```
Comment (hsyuan) [inline src/backend/commands/analyze.c]: Why not just `Ta.%s is NOT NULL` here?

**Pull request merged**